File: Code Generation/responses/python/claude/medium/fruits-into-baskets-iii.py
from typing import List
import sortedcontainers

class Solution:
    def numOfUnplacedFruits(self, fruits: List[int], baskets: List[int]) -> int:
        # Use a sorted structure to efficiently find leftmost basket with capacity >= fruit quantity
        # We need leftmost, so we track (capacity, original_index) and find min index among valid
        
        # Segment tree approach: build over baskets array, query for leftmost index with value >= x
        n = len(baskets)
        
        # Segment tree storing max value in range, to quickly skip ranges
        # For leftmost basket >= fruit, we need to find leftmost position with baskets[pos] >= fruit
        
        size = 1
        while size < n:
            size <<= 1
        
        tree = [0] * (2 * size)
        
        # Build tree with max values
        for i in range(n):
            tree[size + i] = baskets[i]
        for i in range(size - 1, 0, -1):
            tree[i] = max(tree[2*i], tree[2*i+1])
        
        def query_leftmost(val):
            # Find leftmost index in [0, n) where baskets[index] >= val
            if tree[1] < val:
                return -1
            node = 1
            while node < size:
                if tree[2*node] >= val:
                    node = 2*node  # go left
                else:
                    node = 2*node + 1  # go right
            idx = node - size
            if idx < n and tree[node] >= val:
                return idx
            return -1
        
        def update(idx, val):
            pos = size + idx
            tree[pos] = val
            pos >>= 1
            while pos >= 1:
                tree[pos] = max(tree[2*pos], tree[2*pos+1])
                pos >>= 1
        
        unplaced = 0
        for fruit in fruits:
            idx = query_leftmost(fruit)
            if idx == -1:
                unplaced += 1
            else:
                update(idx, 0)  # mark basket as used
        
        return unplaced

File: Code Generation/responses/cpp/claude/medium/fruits-into-baskets-iii.cpp
#include <vector>
#include <climits>
using namespace std;

class SegmentTree {
public:
    static constexpr long long MN = 10000000000LL;
    int n;
    vector<long long> f;
    SegmentTree(int n_) : n(n_), f(4 * n_, MN) {}
    void update(int k, int l, int r, int i, long long x) {
        if (l == r) { f[k] = x; return; }
        int mid = (l + r) / 2;
        if (i <= mid) update(2*k, l, mid, i, x);
        else update(2*k+1, mid+1, r, i, x);
        f[k] = min(f[2*k], f[2*k+1]);
    }
    long long query(int k, int l, int r, int start, int end) {
        if (l == start && r == end) return f[k];
        int mid = (l + r) / 2;
        if (end <= mid) return query(2*k, l, mid, start, end);
        if (mid < start) return query(2*k+1, mid+1, r, start, end);
        long long lp = query(2*k, l, mid, start, mid);
        long long rp = query(2*k+1, mid+1, r, mid+1, end);
        return min(lp, rp);
    }
    // Find leftmost index in [start, end] with value >= minVal
    int findLeftmost(int k, int l, int r, int start, int end, long long minVal) {
        if (f[k] < minVal) return -1;
        if (l == r) return l;
        int mid = (l + r) / 2;
        int res = -1;
        if (start <= mid) {
            res = findLeftmost(2*k, l, mid, start, min(end, mid), minVal);
        }
        if (res == -1 && end > mid) {
            res = findLeftmost(2*k+1, mid+1, r, max(start, mid+1), end, minVal);
        }
        return res;
    }
};

class Solution {
public:
    int numOfUnplacedFruits(vector<int> fruits, vector<int> baskets) {
        int n = fruits.size();
        SegmentTree st(n);
        for (int i = 0; i < n; i++) {
            st.update(1, 0, n-1, i, baskets[i]);
        }
        int unplaced = 0;
        for (int i = 0; i < n; i++) {
            int idx = st.findLeftmost(1, 0, n-1, 0, n-1, (long long)fruits[i]);
            if (idx == -1) {
                unplaced++;
            } else {
                st.update(1, 0, n-1, idx, 0);
            }
        }
        return unplaced;
    }
};